gobblin-config-management/gobblin-config-core/src/main/java/gobblin/config/store/hdfs/SimpleHDFSConfigStoreFactory.java

```diff
@@ -0,0 +1,160 @@
+   */
+  protected Path getDefaultRootDir() throws IOException {
+    return new Path("user", UserGroupInformation.getCurrentUser().getUserName());
+  }
+
+  /**
+   * Gets a default authority if one is not specified. The default authority is the authority of the {@link FileSystem}
+   * the current process is running on. For example, when running on a HDFS node, the authority will taken from the
+   * NameNode {@link URI}.
+   */
+  private String getDefaultAuthority() throws IOException {
+    return FileSystem.get(new Configuration()).getUri().getAuthority();
+  }
+
+  /**
+   * Creates a {@link FileSystem} given a user specified configKey.
+   */
+  private FileSystem createFileSystem(URI configKey) throws ConfigStoreCreationException {
+    try {
+      return FileSystem.get(createHDFSURI(configKey), new Configuration());
+    } catch (IOException | URISyntaxException e) {
+      throw new ConfigStoreCreationException(configKey, e);
+    }
+  }
+
+  /**
+   * Creates a HDFS {@link URI} given a user specified configKey. If the given configKey does not have an authority,
+   * a default one is used instead, provided by the {@link #getDefaultAuthority()} method.
+   */
+  private URI createHDFSURI(URI configKey) throws URISyntaxException, IOException {
+    // Validate the scheme
+    String configKeyScheme = configKey.getScheme();
+    if (!configKeyScheme.startsWith(SIMPLE_HDFS_SCHEME_PREFIX)) {
+      throw new IllegalArgumentException(String.format("Scheme for configKey \"%s\" must begin with \"%s\"!", configKey,
+          SIMPLE_HDFS_SCHEME_PREFIX));
+    }
+
+    if (Strings.isNullOrEmpty(configKey.getAuthority())) {
+      return new URI(getPhysicalScheme(), getDefaultAuthority(), "", "", "");
+    }
+    return new URI(getPhysicalScheme(), configKey.getAuthority(), "", "", "");
+  }
+
+  /**
+   * This method determines the physical location of the {@link SimpleHDFSConfigStore} root directory on HDFS. It does
+   * this by taking the {@link URI} given by the user and back-tracing the path. It checks if each parent directory
+   * contains the folder {@link SimpleHDFSConfigStore#CONFIG_STORE_NAME}. It the assumes this {@link Path} is the root
+   * directory.
+   *
+   * <p>
+   *   If the given configKey does not have an authority, then this method assumes the given {@link URI#getPath()} does
+   *   not contain the dataset root. In which case it uses the {@link #getDefaultRootDir()} as the root directory. If
+   *   the default root dir does not contain the {@link SimpleHDFSConfigStore#CONFIG_STORE_NAME} then a
+   *   {@link ConfigStoreCreationException} is thrown.
+   * </p>
+   */
+  private URI getStoreRoot(FileSystem fs, URI configKey) throws ConfigStoreCreationException {
+    if (Strings.isNullOrEmpty(configKey.getAuthority())) {
+      try {
+        Path defaultRootDir = getDefaultRootDir();
+        if (!fs.exists(new Path(defaultRootDir, SimpleHDFSConfigStore.CONFIG_STORE_NAME))) {
```

Comment:
The ConfigStore assume the CONFIG_STORE_NAME is a directory, we need to add the check in the ConfigStoreFactory ?

That is why I prefer to have CONFIG_STORE_NAME is NOT the parent directory of all the versions